plugins/uefi/efi/fwupdate.c

```diff
@@ -1047,14 +1047,25 @@ add_capsule(update_table *update, EFI_CAPSULE_HEADER **capsule_out,
 	dprint(L"updates guid: %g\n", &update->info->guid);
 	dprint(L"File guid: %g\n", fbuf);
 
+	/*
+	 * A file smaller than 28 bytes is unlikely to be a valid update.
+	 */
+	if (fsize < sizeof(EFI_CAPSULE_HEADER))
+		return EFI_INVALID_PARAMETER;
+
+	capsule = (EFI_CAPSULE_HEADER *)fbuf;
+
 	/*
 	 * See if it has the capsule header, and if not, add one.
 	 *
 	 * Unfortunately there's not a good way to do this, so we're just
 	 * checking if the capsule has the fw_class guid at the right place.
+	 * We also validate CapsuleImageSize in the capsule header in case
+	 * that the update file happens to start with the the fw_class guid.
 	 */
 	if ((guid_cmp(&update->info->guid, (efi_guid_t *)fbuf) == 0 ||
 	     is_fmp_capsule((efi_guid_t *)fbuf)) &&
+	    capsule->CapsuleImageSize == fsize &&
```

Comment:
Are there not capsules with incorrect CapsuleImageSize in the wild? I seem to remember Lenovo might have had some capsules that would have failed this check.

Just so I understand this a bit, a few questions:

* This is for firmware without a capsule header -- can't the vendor just add the header?
* Why are some firmwares using all zeros system GUID?
* Wouldn't that also mean it's zero in the ESRT?
* If it's all zeros in the ESRT, how could we possibly match the correct firmware between models?

Reply:
Something that has bothered me about the fwupd implementation relative to Windows is that appending of the capsule header.  I'm pretty sure it's mandatory for the capsule header on Windows and maybe it's worth drawing that requirement into fwupd too.

Reply:
There's a pretty big OEM shipping firmware without a capsule header. I wonder if it's something we should be checking on the LVFS...

Reply:
And same capsule on Windows?  How can that possibly work with Windows then?

Reply:
I don't know if it's the same capsule. I can certainly run some tests on the existing firmware on the LVFS if that helps. If i'm wrong, and indeed all capsules already have a header then I agree we should drop the header-appending code.

Reply:
Sounds good, will look forward to what you find.

Reply:
tl;dr:

    has_hdr 797
    has_nohdr 148

That's only counting the EFI firmwares on the LVFS. The ones without headers are all Lenovo and NEC. If you'd like the CSV email me, although I'll have to anonymize the filenames for obvious reasons.

Reply:
Well I think the starting point from that is to contact someone at those two companies and ask why no headers? Fwupd bug? Or something else?

If they can change it for the future to have headers that would be ideal.

Reply:
> Are there not capsules with incorrect CapsuleImageSize in the wild? I seem to remember Lenovo might have had some capsules that would have failed this check.
> 
If the capsule header contains incorrect CapsuleImageSize, then it's likely to fail the check in the firmware unless the firmware doesn't implement the update function correctly...

> Just so I understand this a bit, a few questions:
> 
>     * This is for firmware without a capsule header -- can't the vendor just add the header?
> 
>     * Why are some firmwares using all zeros system GUID?
> 
I got the all zeros GUID from two kind of machines.

(1.) The testing system in UEFI Plugfest
One BIOS vendor provided the ROM file starting with the EFI firmware Volume header and somehow their testing system used all zeros GUID as the firmware class GUID, so fwupx64.efi didn't generate the capsule header before UpdateCapsule().

(2.) A production motherboard
This one is the box built by myself recently, and it actually uses the firmware from the vendor of (1.). That means they already shipped all zero GUID firmware in some production systems.

>     * Wouldn't that also mean it's zero in the ESRT?
> 
Yes, the entry in ESRT shows all zeros.

>     * If it's all zeros in the ESRT, how could we possibly match the correct firmware between models?
For LVFS, there should be no firmware with all zeros GUID. I wrote the patch mainly for the testing systems. Besides, it might be useful for those systems with faulty firmware like I have. Of course, the user has to know what she/he is doing...

Reply:
> unless the firmware doesn't implement the update function correctly

No comment there...

> A production motherboard

This is pretty tragic. We simply can't support a zero GUID on production systems; I'd go even further and say we shouldn't create a `FuDevice` for ESRT entries with all zero GUID.

> For LVFS, there should be no firmware with all zeros GUID

I'm hesitant to add support for zero GUIDs to the efi loader. My logic here is that vendors will then say "we'll use a CHID to restrict the update, it's fine" when really it's not fine. I think the un-named vendor needs to stop setting the GUID to zero before it will work on the LVFS and using fwupd.

> The ones without headers are all Lenovo and NEC

Hang on a bit; can't we "fix" up the capsule in Linux userspace before we send the capsule to the loader? I'd be a lot more comfortable doing heuristics like this in the fwupd plugin, rather than in the efi loader.

Reply:
> Hang on a bit; can't we "fix" up the capsule in Linux userspace before we send the capsule to the loader? I'd be a lot more comfortable doing heuristics like this in the fwupd plugin, rather than in the efi loader.

Yes, moving that header appending logic into the OS I think is safer.  We can better record that is happening too then in report metadata.

Reply:
This sounds a better solution. I'll close my pull request.

Reply:
Please review the PR I just opened for that and if you have any comments please share.